module/executiondatasync/optimistic_syncing/pipeline_test.go

```diff
@@ -0,0 +1,460 @@
+// depend on the parent pipeline's state.
+func TestPipelineParentDependentTransitions(t *testing.T) {
+	// Create channels for state updates
+	updateChan := make(chan StateUpdate, 10)
+
+	// Create publisher function
+	publisher := func(update StateUpdate) {
+		updateChan <- update
+	}
+
+	// Create a mock core
+	mockCore := osmock.NewCore(t)
+	mockCore.On("Download", mock.Anything).Return(nil)
+	mockCore.On("Index", mock.Anything).Return(nil)
+	mockCore.On("Persist", mock.Anything).Return(nil)
+
+	// Create a pipeline
+	pipeline := NewPipeline(Config{
+		Logger:               zerolog.Nop(),
+		IsSealed:             false,
+		ExecutionResult:      unittest.ExecutionResultFixture(),
+		Core:                 mockCore,
+		StateUpdatePublisher: publisher,
+	})
+
+	// Start the pipeline
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- pipeline.Run(ctx)
+	}()
+
+	// Initial update - parent in Ready state
+	pipeline.UpdateState(StateUpdate{
+		DescendsFromLastPersistedSealed: true,
+		ParentState:                     StateReady,
+	})
+
+	// Sleep a bit to allow processing
+	time.Sleep(50 * time.Millisecond)
+
+	// Check that pipeline is still in Ready state
+	assert.Equal(t, StateReady, pipeline.GetState(), "Pipeline should remain in Ready state")
+	mockCore.AssertNotCalled(t, "Download")
+
+	// Update parent to downloading
+	pipeline.UpdateState(StateUpdate{
+		DescendsFromLastPersistedSealed: true,
+		ParentState:                     StateDownloading,
+	})
+
+	// Wait for pipeline to progress to WaitingPersist
+	waitForStateUpdate(t, updateChan, StateWaitingPersist)
+	assert.Equal(t, StateWaitingPersist, pipeline.GetState(), "Pipeline should progress to WaitingPersist state")
+	mockCore.AssertCalled(t, "Download", mock.Anything)
+	mockCore.AssertCalled(t, "Index", mock.Anything)
+	mockCore.AssertNotCalled(t, "Persist")
+
+	// Update parent to complete - should allow persisting when sealed
+	pipeline.UpdateState(StateUpdate{
+		DescendsFromLastPersistedSealed: true,
+		ParentState:                     StateComplete,
+	})
+
+	// Mark the execution result as sealed to trigger persisting
+	pipeline.SetSealed()
+
+	// Wait for pipeline to complete
+	waitForStateUpdate(t, updateChan, StateComplete)
+	assert.Equal(t, StateComplete, pipeline.GetState(), "Pipeline should reach Complete state")
+	mockCore.AssertCalled(t, "Persist", mock.Anything)
+
+	// Cancel the context to end the goroutine
+	cancel()
+}
+
+// TestPipelineErrorHandling verifies that errors from Core methods are properly
+// propagated back to the caller.
+func TestPipelineErrorHandling(t *testing.T) {
+	// Test cases for different stages of processing
+	testCases := []struct {
+		name      string
+		setupMock func(mock *osmock.Core)
+	}{
+		{
+			name: "Download Error",
+			setupMock: func(m *osmock.Core) {
+				m.On("Download", mock.Anything).Return(errors.New("download error"))
```

Comment:
how about making these errors variables, then you can check they match with `assert.ErrorIs(t, err, expectedErr)`